swarm/storage/pushsync/pusher_test.go

```diff
@@ -0,0 +1,414 @@
+		cur := int(binary.BigEndian.Uint64(chmsg.Addr[:8]))
+		if cur > max {
+			errf("incorrect order of chunks from db chunk #%d before #%d", cur, max)
+			return nil
+		}
+		v, found := sent.Load(cur)
+		previouslySentAt, repeated := sendTimes[cur]
+		if !found {
+			if !repeated {
+				errf("chunk #%d not sent but received", cur)
+			}
+			return nil
+		}
+		sentAt := v.(time.Time)
+		if repeated {
+			// expect at least retryInterval since previous push
+			if expectedAt := previouslySentAt.Add(retryInterval); expectedAt.After(sentAt) {
+				errf("resync chunk #%d too early. previously sent at %v, next at %v < expected at %v", cur, previouslySentAt, sentAt, expectedAt)
+				return nil
+			}
+		}
+		// remember the latest time sent
+		sendTimes[cur] = sentAt
+		max++
+		// respond ~ mock storer protocol
+		go func() {
+			receipt := &receiptMsg{Addr: chmsg.Addr}
+			rmsg, err := rlp.EncodeToBytes(receipt)
+			if err != nil {
+				errf("error encoding receipt message: %v", err)
+			}
+			log.Debug("chunk sent", "addr", hex.EncodeToString(receipt.Addr))
+			// random delay to allow retries
+			if !delayResponse() {
+				log.Debug("chunk/receipt lost", "addr", hex.EncodeToString(receipt.Addr))
+				return
+			}
+			log.Debug("store chunk,  send receipt", "addr", hex.EncodeToString(receipt.Addr))
+			err = ps.Send(chmsg.Origin, pssReceiptTopic, rmsg)
+			if err != nil {
+				errf("error sending receipt message: %v", err)
+			}
+		}()
+		return nil
+	}
+	// register the respond function
+	ps.Register(pssChunkTopic, false, respond)
+	tags, tagIDs := setupTags(chunkCnt, tagCnt)
+	// construct the mock push sync index iterator
+	tp := newTestPushSyncIndex(chunkCnt, tagIDs, tags, sent)
+	// start push syncing in a go routine
+	p := New(tp, ps, tags)
+	defer p.Close()
+	// collect synced chunks until all chunks synced
+	// wait on errc for errors on any thread
+	// otherwise time out
+	for {
+		select {
+		case i := <-tp.synced:
+			sent.Delete(i)
+			n := synced[i]
+			synced[i] = n + 1
+			if len(synced) == chunkCnt {
+				expTotal := chunkCnt / tagCnt
+				checkTags(t, expTotal, tagIDs[:tagCnt-1], tags)
+				return
+			}
+		case err := <-errc:
+			if err != nil {
+				t.Fatal(err)
+			}
+		case <-time.After(timeout):
+			t.Fatalf("timeout waiting for all chunks to be synced")
+		}
+	}
+
+}
+
+// setupTags constructs tags object create tagCnt - 1 tags
+// the sequential fake chunk i will be tagged with i%tagCnt
+func setupTags(chunkCnt, tagCnt int) (tags *chunk.Tags, tagIDs []uint32) {
+	// construct tags object
+	tags = chunk.NewTags()
+	// all but one tag is created
+	for i := 0; i < tagCnt-1; i++ {
+		tags.New("", int64(chunkCnt/tagCnt))
+	}
+	// extract tag ids
+	tags.Range(func(k, _ interface{}) bool {
+		tagIDs = append(tagIDs, k.(uint32))
+		return true
+	})
+	// add an extra for which no tag exists
+	return tags, append(tagIDs, 0)
+}
+
+func checkTags(t *testing.T, expTotal int, tagIDs []uint32, tags *chunk.Tags) {
```

Comment:
there's already a helper function that does this. see `testutil/tag.go`

Reply:
I'll defer review of this method pending whether you'd replace it with existing helper function.

Reply:
@acud I think we should not have packages in the production code that are used for testing, such as `testutil`. This is also done in the Swarm feeds.

Why exactly can't we have this in a `_test.go` file?

Reply:
i don't mind in which kind of file it is, as long as it is not duplicated functionality